go/worker/compute/executor/committee/proposals.go

```diff
@@ -0,0 +1,82 @@
+package committee
+
+import (
+	"fmt"
+
+	"github.com/google/btree"
+
+	p2pError "github.com/oasisprotocol/oasis-core/go/p2p/error"
+)
+
+// maxPendingProposals is the maximum number of pending proposals that can be queued.
+const maxPendingProposals = 16
+
+// pendingProposals is a priority queue of pending proposals, ordered by round.
+type pendingProposals struct {
+	q *btree.BTreeG[*unresolvedBatch]
+}
+
+func proposalLessFunc(a, b *unresolvedBatch) bool {
+	return a.proposal.Header.Round < b.proposal.Header.Round
+}
+
+func newPendingProposals() *pendingProposals {
+	return &pendingProposals{
+		q: btree.NewG[*unresolvedBatch](2, proposalLessFunc),
+	}
+}
+
+// addPendingProposalLocked adds a new pending proposal that MUST HAVE already undergone basic
+// validity checks and is therefore considered a valid proposal for the given round, but the node's
+// local consensus view may not yet be ready to process the proposal.
+//
+// Must be called with the n.commonNode.CrossNode lock held.
+func (n *Node) addPendingProposalLocked(batch *unresolvedBatch) error {
+	currentRound := n.commonNode.CurrentBlock.Header.Round
+	round := batch.proposal.Header.Round
+
+	// Drop any past proposals.
+	if round <= currentRound {
+		return p2pError.Permanent(fmt.Errorf("proposal round is in the past")) // Do not forward.
+	}
+
+	n.pendingProposals.q.ReplaceOrInsert(batch)
+
+	// In case of overflows, remove the round that is the most in the future.
```

Comment:
We should probably drop past proposals first and if we still have too many proposal, then drop max.

```
for b, ok := n.pendingProposals.q.Min(); ok; {
	if b.proposal.Header.Round >= currentRound {
		break
	}
	n.pendingProposals.q.DeleteMin()
}
```

Reply:
Sounds good. The check probably needs to be `>` though as an equal round is invalid (proposal is always for the next round).